lib_eio_posix/low_level.ml

```diff
@@ -0,0 +1,157 @@
+open Eio.Std
+
+(* There are some things that should be improved here:
+
+   - Blocking FDs (e.g. stdout) wait for the FD to become ready and then do a blocking operation.
+     This might not succeed, and will block the whole domain in that case.
+     Ideally, all blocking operations should happen in a sys-thread instead.
+
+   - Various other operations, such as listing a directory, should also be done in a sys-thread
+     to avoid high latencies in the main domain. *)
+
+type ty = Read | Write
+
+module Fd = Fd
+
+let await_readable fd =
+  Fd.use_exn "await_readable" fd @@ fun fd ->
+  Sched.enter @@ fun t k ->
+  Sched.await_readable t k fd
+
+let await_writable fd =
+  Fd.use_exn "await_writable" fd @@ fun fd ->
+  Sched.enter @@ fun t k ->
+  Sched.await_writable t k fd
+
+let rec do_nonblocking ty fn fd =
+  try fn fd with
+  | Unix.Unix_error (EINTR, _, _) -> do_nonblocking ty fn fd    (* Just in case *)
+  | Unix.Unix_error((EAGAIN | EWOULDBLOCK | EINPROGRESS), _, _) ->
+    Sched.enter (fun t k ->
+        match ty with
+        | Read -> Sched.await_readable t k fd
+        | Write -> Sched.await_writable t k fd
+      );
+    do_nonblocking ty fn fd
+
+let read fd buf start len =
+  if Fd.is_blocking fd then await_readable fd;
+  Fd.use_exn "read" fd @@ fun fd ->
+  do_nonblocking Read (fun fd -> Unix.read fd buf start len) fd
+
+let write fd buf start len =
+  if Fd.is_blocking fd then await_writable fd;
+  Fd.use_exn "write" fd @@ fun fd ->
+  do_nonblocking Write (fun fd -> Unix.write fd buf start len) fd
+
+let sleep_until time =
+  Sched.enter @@ fun t k ->
+  Sched.await_timeout t k time
+
+let socket ~sw socket_domain socket_type protocol =
+  Switch.check sw;
+  let sock_unix = Unix.socket ~cloexec:true socket_domain socket_type protocol in
+  Unix.set_nonblock sock_unix;
+  Fd.of_unix ~sw ~blocking:false ~close_unix:true sock_unix
+
+let rec connect fd addr =
+  try
+    Fd.use_exn "connect" fd (fun fd -> Unix.connect fd addr)
+  with
+  | Unix.Unix_error (EINTR, _, _) -> connect fd addr   (* Just in case *)
+  | Unix.Unix_error ((EAGAIN | EWOULDBLOCK | EALREADY | EINPROGRESS), _, _) ->
+    await_writable fd;
+    match Fd.use_exn "connect" fd Unix.getsockopt_error with
+    | None -> connect fd addr
```

Comment:
This part, on macOS, is failing the networking tests. If I understand it correctly, wouldn't an `EINPROGRESS` plus waiting for the FD to be writeable mean by the time we get to the match statement the error would be cleared and the socket would be connected meaning this second connect would fail with:

```
Eio.Io Unix_error (Socket is already connected, "connect", "")
```

Reply:
You're correct, I think the test is wrong. There shouldn't be a second connect, if SO_ERROR is cleared then it's all done and there's nothing else to do (says Mr.Stevens).
It probably only works on the linux case because a local connection is allowed to never return EINPROGRESS and finish immediately.

Reply:
Makes sense (the man-page is a bit vague about it). What about the other cases (`EAGAIN | EWOULDBLOCK | EALREADY`)?

The Linux man-page says:
> EINPROGRESS The socket is nonblocking and the connection cannot be completed immediately.  (UNIX domain sockets failed with EAGAIN  instead.) [ then talks about waiting and checking for errors ]

But the EAGAIN case doesn't say what you should do in that case.

Reply:
On a TCP socket you shouldn't get EAGAIN/EWOULDBLOCK (IIRC you can get on an AF_UNIX/SOCKSTREAM).

Since now you don't issue a secondary connect you should also never see an EALREADY. 

**But**, I think you're missing the EINTR case, which should be handled the same way as a blocking, which means EINTR does not cancel a connect so you should wait for writeable and check the error.

Reply:
OK. Since we're always nonblocking, I guess we shouldn't normally see EINTR, but I'll treat it the same anyway.